### DISCUSSION TITLE
reportImportCycles and ORM (SqlAlchemy)

### DISCUSSION THREAD

**Discussion by mehdigmira** [microsoft/pyright#4407] [Q&A]
I'm starting to experiment with reportImportCycles. I Think it really can bring value to a codebase.
But we make entensive use of SQLAlchemy, and one of the most common patterns is to have a child reference a parent and the parent itself reference a child.
I've read some of the related issues, and the recommendation seems to be to put both the parent and the child class in the same file.
Though I understand the arguments in favor of doing this, In real life, this would mean putting more than 20 SQLAlchemy models within a single file which would then contains tens of thousands of lines of code. This would be painful to manage.
Do you encounter such cases in your codebases at Microsoft @erictraut ? How do you handle them ?

**Comment by erictraut** [comment]
In my team's codebase, we don't allow import cycles. We also don't use SQLAlchemy in our codebase, so we haven't run into that particular issue. When you say "parent" and "child", are you talking about classes in a class hierarchy, or is it some other child/parent relationship?

**Reply by mehdigmira** [reply]
I'm referring to [ORM relationships](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-one). Typical model declaration in sqlalchemy looks like this

```python
class Parent(Base):
    __tablename__ = "parent_table"
    id = Column(Integer, primary_key=True)
    child_id = Column(Integer, ForeignKey("child_table.id"))
    child: Mapped[Child] = relationship("Child", back_populates="parents")


class Child(Base):
    __tablename__ = "child_table"
    id = Column(Integer, primary_key=True)
    parents: Mapped[list[Parent]] = relationship("Parent", back_populates="child")
```

See how you can call parent.child, but also child.parents. Of course this is a minimal example. In real life, Child itself would have other relationship and Parent would have multiple child tables. 
So Solving this without using TYPE_CHECKING requires:
- everytime there is a 2-way relationship between models, put them together in the same file. This leads to huge files.
- destroying these 2-way relationship to only have a 1-way relationship. In some cases this can be done, but many times the 2-way relationship is very useful

**Reply by erictraut** [reply]
Yeah, if classes are mutually dependent like this, you would need to either define them in the same module or live with import cycles. Perhaps there are subsets of your models that are independent and could be grouped into some number of separate modules without any generating import cycles.

**Reply by mehdigmira** [reply]
I'm already trying to split the classes by dependency, but is still end up with some of the modules being huge.
I'm thinking of the following approach:
- if the number of dependencies is reasonable, put them in the same module
- if not create a folder with modules that will have circular dependencies in between each other, and disable reportImportCycles on each of these modules
It's not a perfect solution, but maybe it's the least evil ^^

**Reply by mehdigmira** [reply]
just an additional question, if i have the following structure:
- module1: reportImportCycles=false
- module2: reportImportCycles=true
would a a circular dependency between module1 and module2 be detected ? (I hope so, but not sure, since I've read that the module on which the circular dependency is logged is somewhat "random")

**Reply by erictraut** [reply]
You cannot enable `reportImportCycles` on a module-by-module basis. It's a diagnostic check that is performed across modules, not within modules.

**Reply by mehdigmira** [reply]
In https://github.com/microsoft/pyright/issues/3489#issuecomment-1133496121 you stated that "you can suppress the reportImportCycles output for a specific file using a # pyright: reportImportCycles=false comment within that file". That's a way to silence the error on a given module right ?

**Reply by mehdigmira** [reply]
Basically what i am trying to do

```python
# model1.py

# I know this contains sqlalchemy models and i cannot avoid having circular dependencies with other models
# so i'm just disabling it on every such module, so that i'm sure it will never be reported
# pyright: reportImportCycles=false

```

```python
# model2.py

# I know this contains sqlalchemy models and i cannot avoid having circular dependencies with other models
# so i'm just disabling it on every such module, so that i'm sure it will never be reported
# pyright: reportImportCycles=false

```

```python
# controller.py

# Now this is a controller, it should not have circular dependencies with anything else
# I hope that by setting the flag here any circular dependency between controller.py
# and model1.py or model2.py will be reported
# pyright: reportImportCycles=true

```

**Reply by erictraut** [reply]
The check is performed across modules. If a cycle is found, a diagnostic is generated for one of the files in the cycle (in no particular guaranteed order). You can suppress that diagnostic in that particular file by adding `# pyright: reportImportCycles=false`. If you change your code or we slightly tweak the code in pyright, the diagnostic could be reported against a different file in the cycle. So this could end up being a somewhat fragile approach.

**Reply by mehdigmira** [reply]
Yes, I understand that, sorry if i don't express myself well, but my question is:
"If i have a circular dependency between M1 (module with reportImportCycles=false) and M2 (module with reportImportCycles=true), will pyright **always** report the issue on M2 or is there a chance this will be silenced because pyright will try to report it on M1 which silences this issue ?"

(Thank you for your time)

**Reply by erictraut** [reply]
If you have a circular dependency between two modules M1 and M2, pyright will detect the cycle and report an error against either M1 or M2, not both. You don't have any control over which one it will choose. It's dependent on the order in which files are analyzed and the internal algorithm used. The feature isn't designed to be used for detecting some cycles and not others. You may find that the module that pyright uses to report the cycle is sufficiently deterministic that you can use this approach, but it is fragile and could easily change due to changes in the algorithm or changes to your code (e.g. the addition of a new import statement in one of the modules that causes them to be analyzed in a different order).

**Reply by mehdigmira** [reply]
would it be possible to change the algorithm to always report the error against one member of the cycle that doesn't have reportImportCycles=false ?

This would leave the logic of how cycles are detected, and what is detected as a cycle untouched, but still offer the possibility to exclude some known cycles by putting a reportImportCycles=false on every single module of the cycle, while still having a guarantee that cycles outside of this one will be reported.

I know that cycles should be avoided, but I can assure that the best anyone using an ORM on a large codebase can do is to isolate known cycles between known classes. I feel like this would be a good compromise, don't you think ?

I can try to prepare a PR if you're ok with this

**Reply by erictraut** [reply]
No, the code for detecting and reporting cycles is completely separate from the code for suppressing file-level diagnostics. They have nothing to do with each other. Sorry, but you're attempting to use this feature in a way that it wasn't designed.

**Reply by erictraut** [reply]
I think you need to just recognize that your codebase contains import cycles, so pyright's import cycle detection probably won't work for you. Perhaps there are other linters that can help you out here if you are really intent on detecting and preventing some (but not all) import cycles in your code.

**Reply by mehdigmira** [reply]
Alright, then I guess reporting the error on every module of the cycle wouldn't be an option either ? ^^
Too bad, this makes this option nearly unusable by codebases that rely on ORMs, I guess I'll to live with that ! 😛
